### PR TITLE
cli: Enhance tooling for checking dependencies

### PR DESCRIPTION
**Pull request by planger**
#### What it does

Enhances the `theia` CLI with dedicated checks regarding pulled dependencies as detailed below.

##### `theia check:theia-version`

*Check that all dependencies have been resolved to the same Theia version.*

It'll obtain all package workspaces of the monorepo from `package.json/workspaces` and run through their `node_modules` as well as the root `node_modules` in order to collect all dependencies to `@theia/**` packages and test whether they all have the same version. It'll exclude `@theia/monaco-editor-core` though, as it follows the version of Monaco, which differs from the common Theia version.

Parameters:
* `--suppress` or `-s`: Suppress exiting with failure code

##### `theia check:hoisted`

*Check that all dependencies are hoisted.*

This command behaves as before, but uses the same enhanced implementation as the command above and below. Also the output is a bit improved with respect to readability in comparison to the current implementation.

##### `theia check:dependencies`

*Check uniqueness of dependency versions or whether they are hoisted.*

This command is the basis for the two commands above, but exposes the full configurability to adjust the dependency checks to specific needs.

Parameters:
* ` --workspaces` or `-w`: Glob patterns of workspaces to analyze, relative to `cwd`. [array] [default: All glob patterns listed in the package.json's workspaces]
* `--include` or `-i`: Glob pattern of dependencies' package names to be included, e.g. `-i "@theia/**"`. [array] [default: ["**"]]
* `--exclude` or `-e`: Glob pattern of dependencies' package names to be excluded. [array] [default: None]
* `--skip-hoisted` or `-h`: Skip checking whether dependencies are hoisted. [boolean] [default: false]
* `--skip-uniqueness` or `-u`: Skip checking whether all dependencies are resolved to a unique version. [boolean] [default: false]
* `--skip-single-theia-version` or `-t`: Skip checking whether all @theia/** dependencies are resolved to a single version. [boolean] [default: false]
* `--suppress` or `-s`: Suppress exiting with failure code. [boolean] [default: false]

See https://github.com/eclipse-theia/theia/issues/11482 for the problem this PR aims to address.

Contributed on behalf of STMicroelectronics.

#### How to test

There are three commands (see above) to test.
For `theia check:hoisted` it should be tested whether it prints the same amount of information as before. However, the content may look slightly different now:

image

To test `yarn check:theia-version` you could proceed as follows:

* Check out this PR and build (`yarn`)
* Go to `dev-packages/cli` and run `yarn link`
* Create a sample Theia application, e.g. with the theia generator, and run `yarn link "@theia/cli"` in its root

* Running `yarn && yarn theia check:theia-version` should give no errors: 
image

* To test different Theia versions of non-overlapping dependencies, you could remove `@theia/terminal` from the dependencies of `browser-app` and `electron-app`, set all their dependencies to `1.28.0`, but add `@theia/terminal` as a dependency _with a `1.26.0`_ in a Theia extension of your project. Note how no `@theia/*` dependencies are hoisted. However, in the root `node_modules` you'll find a mix of versions in the `@theia/*` dependencies (terminal vs the others). Now run `yarn && yarn theia check:theia-version` again and it should print: 
image

* You can now re-add `@theia/terminal` with version `1.28.0` to `browser-app`, which will also lead to non-hoisted dependencies and re-run `yarn && yarn theia check:theia-version -s` (suppress). You'll get now one error (the same error as above), but also a warning that there is a dependency that isn't hoisted. Also note how the exit code differs because of the `-s`: 
image

To test `yarn check:dependencies` you could proceed as above, but now play around with the parameters. Implementation-wise it is the same as the two commands above. Try e.g.
`yarn theia check:dependencies --workspaces "packages/**" --exclude "rimraf/**" --skip-single-theia-version`
which will find non-hoisted dependencies, only the packages of the folder `packages`, but will not show an error for `rimraf`.

Don't forget to `yarn unlink` in `dev-packages/cli` after testing.

#### Review checklist

- [X] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by planger** [comment]
Just wanted to ping if someone would be available to give feedback to this proposal? :relaxed: Thanks a lot!

**Comment by planger** [comment]
@msujew Did you have a chance to look at this PR? I'm looking forward to your feedback. Thanks a lot!

File: dev-packages/cli/src/check-dependencies.ts [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    const remainingIndent = indent - issueNumber.toString().length;
+    const indentString = ' '.repeat(remainingIndent + 1);
+    console.log(issueTitle(issue, issueNumber, indentString));
+    console.log(issueDetails(issue, '   ' + ' '.repeat(indent)));
+    console.log();
+}
+
+function issueTitle(issue: DependencyIssue, issueNumber: number, indent: string): string {
+    const dependent = issue.package.dependent ? ` in ${chalk.blueBright(issue.package.dependent ?? 'unknown')}` : '';
+    return chalk.bgWhiteBright.bold(`#${issueNumber}${indent}`) + ' ' + chalk.cyanBright(issue.package.name)
```
Comment (vince-fugnitto) [inline dev-packages/cli/src/check-dependencies.ts:269]: I'm not sure if it is my setup or terminal preferences but I do not see the text (white on white) for the issue number:

Screen Shot 2022-08-23 at 1 59 49 PM

File: dev-packages/cli/src/check-dependencies.ts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    for (const include of options.include) {
+        matchingPackageJsons.push(
+            ...glob.sync(include + '/' + PACKAGE_JSON, {
+                cwd: nodeModulesDir,
+                ignore: [`**/${NODE_MODULES}/**`, ...options.exclude]
+            })
+        );
+    }
+    return matchingPackageJsons.map(packageJson => toDependency(packageJson, nodeModulesDir, dependent));
+}
```
Comment (paul-marechal) [inline dev-packages/cli/src/check-dependencies.ts:126]: Please consider the following code that avoids iterating twice over everything:

```suggestion
function findDependencies(workspace: string, options: CheckDependenciesOptions): Package[] {
    const dependent = getPackageName(path.join(process.cwd(), workspace, PACKAGE_JSON));
    const nodeModulesDir = path.join(workspace, NODE_MODULES);
    const matchingPackageJsons: Package[] = [];
    options.include.forEach(include => {
        glob.sync(`${include}/${PACKAGE_JSON}`, {
            cwd: nodeModulesDir,
            ignore: [`**/${NODE_MODULES}/**`, ...options.exclude]
        })
        .forEach(packageJson => {
            matchingPackageJsons.push(toDependency(packageJson, nodeModulesDir, dependent));
        });
    });
    return matchingPackageJsons;
}
```

File: dev-packages/cli/src/check-dependencies.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+function deriveWorkspaces(options: CheckDependenciesOptions): string[] {
+    const wsGlobs = options.workspaces ?? readWorkspaceGlobsFromPackageJson();
+    const workspaces: string[] = [];
+    for (const wsGlob of wsGlobs) {
+        workspaces.push(...glob.sync(wsGlob + '/'));
+    }
+    return workspaces;
+}
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply dev-packages/cli/src/check-dependencies.ts:92]: I believe that the use of `workspaces` downstream should be optional indeed.

[the remark this answers is not in the document]
Comment (planger) [reply dev-packages/cli/src/check-dependencies.ts:92]: Great suggestion, thanks! I'll make it optional.
I indeed only tested a project without `workspaces` in combination with the option `--workspaces`.

File: dev-packages/cli/src/check-dependencies.ts (hunk 1 above)
Comment (planger) [reply dev-packages/cli/src/check-dependencies.ts:269]: Hm, interesting. I'll amend a change to set the font to black explicitly. Can you please test with your theme and terminal preferences, whether this works for you?

File: dev-packages/cli/src/check-dependencies.ts (hunk 2 above)
Comment (planger) [reply dev-packages/cli/src/check-dependencies.ts:126]: Sure, thanks, that is probably more efficient!

**Comment by planger** [comment]
@vince-fugnitto @paul-marechal Thanks a lot for your feedback! I pushed an update which includes your suggestions.

**Review by vince-fugnitto — COMMENTED** [review]
I confirm that the changes work better 👍 
I'll give @paul-marechal a chance to perform a final review as well.

File: dev-packages/cli/package.json [hunk 4]
```diff
@@ -47,7 +47,9 @@
     "puppeteer": "^2.0.0",
     "puppeteer-to-istanbul": "^1.2.2",
     "temp": "^0.9.1",
-    "yargs": "^15.3.1"
+    "yargs": "^15.3.1",
+    "glob": "^8.0.3",
+    "log-update": "^4.0.0"
```
[the remark this answers is not in the document]
Comment (planger) [reply dev-packages/cli/package.json]: Sure, no problem! Done with https://github.com/eclipse-theia/theia/commit/94ac97460ee55464ed9c642995c10d573725cf7b

**Comment by paul-marechal** [comment]
Using the following `package.json`

```json
{
  "name": "theia-app-example",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "@theia/core": "^1.28.0",
    "@theia/plugin-ext": "^1.28.0",
    "@theia/plugin-ext-vscode": "^1.28.0"
  },
  "devDependencies": {
    "@theia/cli": "^1.28.0"
  }
}
```

Running `theia check:dependencies` seems to find errors that I believe shouldn't be picked up:

image

Notice what the paths look like: They aren't nested dependencies... As a rule of thumb I think we should only consider packages like `node_modules/package.json`.

**Comment by paul-marechal** [comment]
Otherwise I tried polluting my example app by adding `@theia/workspace@1.25.0` to force a duplication and it was correctly picked up. I think it should be good to merge once my previous issue is fixed.

**Comment by planger** [comment]
Thanks again for your feedback!
@paul-marechal You're right, we should filter package.json's deeply nested in dependencies. I've now added these two ignore globs:
```javascript
`[^@]*/*/**/${PACKAGE_JSON}`, // package.json that isn't at the package root (and not in an @org)
`@*/*/*/**/${PACKAGE_JSON}`, // package.json that isn't at the package root (and in an @org)
```

I confirmed that the dependencies you mentioned aren't showing up, while other expected ones still show up.
Can you confirm this works for your example too?

**Review by paul-marechal — APPROVED** [review]
Successfully picks up duplicated Theia package version, this will be helpful!

**Pull request merged**